activerecord/test/support/method_stub.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module MethodStub
+  private
+
+    def with_stubbed_method(object, method_name, return_value = nil)
+      stub_method(object, method_name, return_value)
+
+      yield
+
+    ensure
+      unstub_method(object, method_name)
+    end
+
+    # TODO: `alias_method` is private in Ruby 2.4, so once the minimum
+    # version is 2.5 there'll be no need to resort to `module_eval`.
+
+    def stub_method(object, method_name, return_value = nil)
+      base_var_name = method_name.to_s.gsub(/[?!]/, "")
+
+      object.class_eval(<<-EOF, __FILE__, __LINE__ + 1)
+        eval "@@#{base_var_name}_called = 0"
+        eval "@@#{base_var_name}_args = []"
+
+        define_method("stubbed_#{method_name}") do |*args|
+          eval "@@#{base_var_name}_called += 1"
+          eval "@@#{base_var_name}_args += args"
```

Comment:
How else will we define these class variables dynamically? With `send` ?

Reply:
I guess using of just `@@#{base_var_name}_called += 1` should work as well.
And what about using of instance variabes instead of class variables then we are able to use `remove_instance_variable` http://ruby-doc.org/core-2.5.1/Object.html#method-i-remove_instance_variable

Reply:
You are right! Removing the redundant `eval`s.